### PR TITLE
Initial CAST support for MATLAB

### PR DESCRIPTION
**Pull request by jastier**
This PR adds an early MATLAB parser to the main branch CAST languages.  (PR #382)
Pytest integration is tested by this CI.
Test cases are parsed from MATLAB source to GroMEt wiring output.

### CODE REVIEW

File: Dockerfile.mathjax
Comment (myedibleenso) [inline Dockerfile.mathjax:1]: @jastier , please remove this file.

File: skema/program_analysis/CAST/cast_walker.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    ModelReturn,
+    ModelIf,
+    RecordDef,
+    Attribute,
+)
+
+
+# Generate a CAST for matlab and visit every node
+
+LANGUAGE_LIBRARY_REL_PATH = 'tree-sitter-matlab' # TODO derive
```
Comment (myedibleenso) [inline skema/program_analysis/CAST/cast_walker.py:33]: @vincentraymond-ua , how do you recommend retrieving this?

File: skema/program_analysis/CAST/matlab/README.md
Comment (myedibleenso) [inline skema/program_analysis/CAST/matlab/README.md:1]: this README mentions several scripts.  Are these included here?  If not we should either update the README or add versions of those scripts.

File: skema/program_analysis/CAST/matlab/tests/data/assignment.m
Comment (myedibleenso) [inline skema/program_analysis/CAST/matlab/tests/data/assignment.m:1]: Can these files be moved into `data/gromet/matlab` and imported in your tests?

File: skema/program_analysis/CAST/matlab/__init__.py
Comment (myedibleenso) [inline skema/program_analysis/CAST/matlab/__init__.py:1]: This relates to how we generate documentation.  In `skema/program_analysis/CAST/matlab/__init__.py`, please try adding the following:

```python
__pdoc__ = {
    'tests': False,
    'data': False
}
```

**Review by myedibleenso — CHANGES_REQUESTED** [review]
Thanks, @jastier . 

Regarding the doc generation error, can you please try the following?  Within `skema/program_analysis/__init__.py`, please try adding the following:

```python
__pdoc__ = {
    'tree_sitter_parsers': False
}
```

File: skema/program_analysis/CAST/cast_walker.py (hunk 1 above)
Comment (vincentraymond-ua) [reply skema/program_analysis/CAST/cast_walker.py:33]: You can import the shared object file path directly from build parsers and use it like so:

1. Replace line 33  with `from skema.program_analysis.tree_sitter_parsers.build_parsers import INSTALLED_LANGUAGES_FILEPATH`
2. Replace line 46 with `INSTALLED_LANGUAGES_FILEPATH`

File: skema/program_analysis/CAST/fortran/.gitignore
Comment (vincentraymond-ua) [inline skema/program_analysis/CAST/fortran/.gitignore:1]: Could we point this to ignore everything in `skema/program_analysis/tree_sitter_parsers/build`?

File: skema/program_analysis/CAST/matlab/README.md
Comment (vincentraymond-ua) [reply skema/program_analysis/CAST/matlab/README.md:1]: I think this is left over from when the grammar was stored in the repository instead of cloned at run time.

File: skema/program_analysis/CAST/matlab/matlab_tree_builder.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
+import json
+import os.path
+import pprint
+from pathlib import Path
+from typing import Any, Dict, List, Union
+from tree_sitter import Language, Parser, Node, Tree
+
+from skema.program_analysis.tree_sitter_parsers.build_parsers import INSTALLED_LANGUAGES_FILEPATH
+
+class MATLAB_TREE_BUILDER(object):
```
Comment (vincentraymond-ua) [inline skema/program_analysis/CAST/matlab/matlab_tree_builder.py:10]: This isn't a major issue, but generally class names in Python are PascalCase (MATLAB_TREE_BUILDER -> MatlabTreeBuilder)

File: skema/program_analysis/CAST/matlab/matlab_tree_builder.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    def __init__(self):
+        
+        # Create the tree-sitter MATLAB parser
+        self.parser = Parser()
+        self.parser.set_language(
+            Language(
+                Path(Path(__file__).parent, INSTALLED_LANGUAGES_FILEPATH),
```
Comment (vincentraymond-ua) [inline skema/program_analysis/CAST/matlab/matlab_tree_builder.py:17]: This can now be simplified to just `INSTALLED_LANGUAGES_FILEPATH`

File: skema/program_analysis/CAST/matlab/node_helper.py [hunk 4]
```diff
@@ -0,0 +1,111 @@
+from typing import List, Dict
+from skema.program_analysis.CAST2FN.model.cast import SourceRef
+
+from tree_sitter import Node
+
+CONTROL_CHARACTERS = [
```
Comment (vincentraymond-ua) [inline skema/program_analysis/CAST/matlab/node_helper.py:6]: The set of control characters may be different between Fortran and Matlab. This isn't something that needs to be changed for this PR, but is something to be aware might cause issues down the road.

File: skema/program_analysis/CAST/matlab/tests/test_matlab_parse_tree.py [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+from pathlib import Path
+
+from tree_sitter import Language, Parser
+
+# test parser
+from skema.program_analysis.CAST.matlab.matlab_to_cast import MATLAB2CAST
+
+# test for existence of shared grammar object
+from skema.program_analysis.tree_sitter_parsers.build_parsers import INSTALLED_LANGUAGES_FILEPATH
+if os.path.exists(INSTALLED_LANGUAGES_FILEPATH):
```
Comment (vincentraymond-ua) [inline skema/program_analysis/CAST/matlab/tests/test_matlab_parse_tree.py:12]: INSTALLED_LANGUAGES_FILEPATH is a pathlib.Path object, so you should be able to do:
`INSTALLED_LANGUAGES_FILEPATH.exists()`

File: skema/program_analysis/CAST/matlab/tests/test_matlab_parse_tree.py [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    print('Shared object at: ' + str(INSTALLED_LANGUAGES_FILEPATH))
+else:  
+    # If not found, create it
+    print('Did not find shared object at ' + str(INSTALLED_LANGUAGES_FILEPATH))
+    from skema.program_analysis.tree_sitter_parsers.build_parsers import build_parsers
+    build_parsers('--matlab')
```
Comment (vincentraymond-ua) [inline skema/program_analysis/CAST/matlab/tests/test_matlab_parse_tree.py:18]: Ignore if already working, but this may need to `build_parsers(["matlab"])`

File: skema/program_analysis/CAST2FN/ann_cast/cast_to_annotated_cast.py [hunk 7]
```diff
@@ -62,8 +62,8 @@ def generate_annotated_cast(self, grfn_2_2: bool = False):
     def visit(self, node: AstNode) -> AnnCastNode:
         # print current node being visited.
         # this can be useful for debugging
-        # class_name = node.__class__.__name__
-        # print(f"\nProcessing node type {class_name}")
+        class_name = node.__class__.__name__
+        print(f"\nProcessing node type {class_name}")
```
Comment (vincentraymond-ua) [inline skema/program_analysis/CAST2FN/ann_cast/cast_to_annotated_cast.py:66]: Lets re-comment these debugging lines

File: skema/program_analysis/single_file_ingester.py [hunk 8]
```diff
@@ -3,6 +3,11 @@
 import os
 from pathlib import Path
 
+# Add the execution directory to path so skema imports work
```
Comment (vincentraymond-ua) [inline skema/program_analysis/single_file_ingester.py:6]: Could you clarify what this is doing?

Comment (titomeister) [reply skema/program_analysis/single_file_ingester.py:6]: I think this is adding the paths of the skema environment into the current system? That way the 'skema....' imports work properly.
I think this shouldn't be needed if the skema virtual environment is installed properly. I've never had to do this before.

**Review by titomeister — COMMENTED** [review]
I didn't see anything that looked off to me. I think Vincent's comments are pretty on point especially since he's worked with tree sitter more. I think following his style for generating CAST is good, that's what I plan on doing as well.

File: skema/program_analysis/CAST/matlab/node_helper.py (hunk 4 above)
Comment (jastier) [reply skema/program_analysis/CAST/matlab/node_helper.py:6]: Yes, same goes for a whole lot of reserved words and functions.   It's what I'm working on now as I build out my test cases.

File: skema/program_analysis/CAST/matlab/tests/test_matlab_parse_tree.py (hunk 6 above)
Comment (jastier) [reply skema/program_analysis/CAST/matlab/tests/test_matlab_parse_tree.py:18]: Implemented.   Good call.

File: skema/program_analysis/CAST/matlab/tests/test_matlab_parse_tree.py (hunk 5 above)
Comment (jastier) [reply skema/program_analysis/CAST/matlab/tests/test_matlab_parse_tree.py]: Done.

File: skema/program_analysis/CAST/matlab/__init__.py
Comment (jastier) [reply skema/program_analysis/CAST/matlab/__init__.py:1]: Done.

File: skema/program_analysis/CAST/matlab/matlab_tree_builder.py (hunk 2 above)
Comment (jastier) [reply skema/program_analysis/CAST/matlab/matlab_tree_builder.py]: Absolutely.   Done.

File: skema/program_analysis/CAST/matlab/tests/data/assignment.m
Comment (jastier) [reply skema/program_analysis/CAST/matlab/tests/data/assignment.m:1]: Done.

File: Dockerfile.mathjax
Comment (jastier) [reply Dockerfile.mathjax:1]: Done.

File: skema/program_analysis/CAST/cast_walker.py (hunk 1 above)
Comment (jastier) [reply skema/program_analysis/CAST/cast_walker.py]: That file has been deleted, the functionality was subsumed into CAST/matlab/matlab_to_cast.py

File: data/gromet/matlab/assignment_1/single_line_assignment.m
Comment (myedibleenso) [inline data/gromet/matlab/assignment_1/single_line_assignment.m:1]: Sorry to go back and forth on this.  Given what we're doing in #494 for #495 , let's move this under `skema/program_analysis/CAST/matlab/tests/data/`

File: data/gromet/matlab/assignment_2/multi_line_assignment.m
Comment (myedibleenso) [inline data/gromet/matlab/assignment_2/multi_line_assignment.m:1]: Sorry to go back and forth on this.  Given what we're doing in #494 for #495 , let's move this under `skema/program_analysis/CAST/matlab/tests/data/`

File: skema/program_analysis/CAST/matlab/matlab_tree_builder.py (hunk 3 above)
Comment (myedibleenso) [reply skema/program_analysis/CAST/matlab/matlab_tree_builder.py:18]: @vincentraymond-ua , to clarify, are you saying that `Path(Path(__file__).parent, INSTALLED_LANGUAGES_FILEPATH)` can be replaced with just `INSTALLED_LANGUAGES_FILEPATH`?

File: skema/program_analysis/single_file_ingester.py (hunk 8 above)
Comment (myedibleenso) [reply skema/program_analysis/single_file_ingester.py:6]: Agreed. @jastier , this shouldn't be necessary if the module installed correctly.

File: data/gromet/matlab/assignment_2/multi_line_assignment.m
Comment (jastier) [reply data/gromet/matlab/assignment_2/multi_line_assignment.m:1]: Not a problem at all.   Done.

File: skema/program_analysis/CAST/matlab/README.md
Comment (jastier) [reply skema/program_analysis/CAST/matlab/README.md:1]: @vincentraymond-ua   Correct.
The file has been deleted.

File: skema/program_analysis/CAST2FN/ann_cast/cast_to_annotated_cast.py (hunk 7 above)
Comment (jastier) [reply skema/program_analysis/CAST2FN/ann_cast/cast_to_annotated_cast.py]: Done.

File: skema/program_analysis/CAST/matlab/__init__.py
Comment (myedibleenso) [inline skema/program_analysis/CAST/matlab/__init__.py:1]: This one relates to the API documentation generation.  Please try adding the following to this file:

```python
__pdoc__ = {
    'tests': False
}
```

File: skema/program_analysis/CAST/matlab/matlab_tree_builder.py (hunk 3 above)
Comment (myedibleenso) [reply skema/program_analysis/CAST/matlab/matlab_tree_builder.py:18]: @jastier , looking at the definition of `INSTALLED_LANGUAGES_FILEPATH`, you should be able to replace `Path(Path(__file__).parent, INSTALLED_LANGUAGES_FILEPATH)` with just `INSTALLED_LANGUAGES_FILEPATH`

Comment (jastier) [reply skema/program_analysis/CAST/matlab/matlab_tree_builder.py]: Done.

File: skema/program_analysis/single_file_ingester.py (hunk 8 above)
Comment (jastier) [reply skema/program_analysis/single_file_ingester.py]: It is no longer needed and has been removed.

File: skema/program_analysis/CAST/fortran/.gitignore
Comment (jastier) [reply skema/program_analysis/CAST/fortran/.gitignore:1]: Done.   

- The .gitignore files in `skema/program_analysis/CAST/matlab` and `skema/program_analysis/CAST/fortran` have been removed.
- A new .gitignore has been added in `skema/program_analysis/tree_sitter_parsers` to ignore the build directory.

File: skema/program_analysis/CAST/matlab/__init__.py
Comment (jastier) [reply skema/program_analysis/CAST/matlab/__init__.py:1]: Done.

**Comment by myedibleenso** [comment]
@jastier , it looks like one of the changes in this PR has broken the installation: https://github.com/ml4ai/skema/actions/runs/6242837244/job/16947373030#step:12:36

I'm not sure where the issue is at the moment, but you should be able to debug this locally by following the dev environment setup instructions on the project website.

File: __init__.py
Comment (myedibleenso) [inline __init__.py:1]: This needs to be removed.  Why was this added, @jastier ? There isn't a module at that level...

Comment (jastier) [reply __init__.py:1]: Done.

**Review by myedibleenso — APPROVED** [review]

**Comment by myedibleenso** [comment]
Thanks, @jastier
